lunabot_control/scripts/homing_controller.py

```diff
@@ -0,0 +1,361 @@
+
+
+class State(Enum):
+    DEPOSIT = 0
+    DRIVE = 1
+    HOMING = 2
+    RESET = 3
+    DONE = 4
+    RETRACT = 5
+    OPEN_EXC = 6
+    CLOSE_EXC = 7
+
+
+class HomingController:
+    # clips input is within limits bounds
+    linear_threshold = 0.05
+    angular_threshold = 3 * np.pi / 180
+
+    linear_setpoint = 0.8
+    angular_setpoint = 0
+
+    KP = np.array([4, 4])
+    KI = np.array([0.0, 0.0])
+    KD = np.array([0, 0.0])
+
+    ALPHA = 0.5  # diff drive
+    SCALE = 0.5
+    MIN_VEL = 0.2
+
+    DRIVE_CTRL = 0.5
+    DRIVE_CURR_RANGE = (5000, 18000)
+    DEP_CURR_RANGE = (12900, 13400)  # TODO: find
+
+    TIME_FROM_RETRACT_TO_EXTEND = 25.0
+    TIME_TO_CLEAR_BUCKETS = 7
+    DEP_CURR_STARTING = 2
+
+    def __init__(self):
+        self._apriltag_sub = rospy.Subscriber(
+            "/d455_front/camera/color/tag_detections",
+            AprilTagDetectionArray,
+            self.apritag_cb,
+        )
+
+        self._state_sub = rospy.Subscriber("/state", RobotState, self._robot_state_cb)
+        self._effort_pub = rospy.Publisher("/effort", RobotEffort, queue_size=1)
+
+        self._effort_msg = RobotEffort()
+        self._state = State.OPEN_EXC
+
+        self._curr_time_act = None
+
+        self.T_camera_april_msg = None
+        # self._effort_msg = Twist()
+        # self._effort_msg = Float32()
+        self._prev_error = np.zeros(2)
+        self._curr_error = np.zeros(2)
+        self._error_total = np.zeros(2)
+
+        r = rospy.Rate(20)
+
+        rospy.on_shutdown(self.shutdown_hook)
+
+        while not rospy.is_shutdown():
+            if (
+                self._drive_left_curr is not None
+                and self._drive_right_curr is not None
+                and self._dep_curr is not None
+            ):
+                if self._state == State.DONE:
+                    break
+                else:
+                    self.loop()
+            r.sleep()
+
+    def shutdown_hook(self):
+        self.stop()
+        print("stopping homing control")
+
+    def apritag_cb(self, msg):
+        if len(msg.detections) != 0:
+            # self._effort_msg.linear.x = 0
+            # self._effort_msg.angular.z = 0
+
+            self.T_camera_april_msg = msg.detections[
+                0
+            ].pose.pose.pose  # transformation from apriltag to camera frame (check header frame_id to be sure)
+        else:
+            self.T_camera_april_msg = None
+
+    def _robot_state_cb(self, msg):
+        self._drive_left_curr = msg.drive_left_curr
+        self._drive_right_curr = msg.drive_right_curr
+        self._dep_curr = msg.dep_curr
+
+    def homing(self):
+        if self.T_camera_april_msg is None:
+            return
+        T_camera_april = ros_numpy.numpify(
+            self.T_camera_april_msg
+        )  # Converting the transformation matrix to a numpy array
+
+        print(T_camera_april)
+
+        # Assigning the unit vector values for the different vectors
```

Comment:
Maybe add a more detialed comment explaning the math (as we got confused by this several times)

Reply:
leaving this for a future docs addition with mkdocs